### PR TITLE
[RLlib] Fix issue (bug): LSTM + non-shared vf + PPO + tuple actions

### LINKED ISSUE #6884: [RLlib] Using LSTM model raises ValueError (Tuple obs_space, similar to #3367)
### System Information
- Fedora 7.7 (Maipo) server
- Ray from pip
- Ray version 0.8.0
- Python version 3.6.8
### Problem
I have a custom environment with a Tuple observation space, using the default model in rllib that is trained using PPO.  Everything works when I don't turn on the LSTM option.  However, when I do turn it on, I get the following stack trace:
`
File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/agents/trainer_template.py", line 90, in __init__
    Trainer.__init__(self, config, env, logger_creator)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/agents/trainer.py", line 398, in __init__
    Trainable.__init__(self, config, logger_creator)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/tune/trainable.py", line 96, in __init__
    self._setup(copy.deepcopy(self.config))
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/agents/trainer.py", line 523, in _setup
    self._init(self.config, self.env_creator)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/agents/trainer_template.py", line 109, in _init
    self.config["num_workers"])
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/agents/trainer.py", line 568, in _make_workers
    logdir=self.logdir)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/evaluation/worker_set.py", line 64, in __init__
    RolloutWorker, env_creator, policy, 0, self._local_config)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/evaluation/worker_set.py", line 220, in _make_worker
    _fake_sampler=config.get("_fake_sampler", False))
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/evaluation/rollout_worker.py", line 350, in __init__
    self._build_policy_map(policy_dict, policy_config)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/evaluation/rollout_worker.py", line 766, in _build_policy_map
    policy_map[name] = cls(obs_space, act_space, merged_conf)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/policy/tf_policy_template.py", line 143, in __init__
    obs_include_prev_action_reward=obs_include_prev_action_reward)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/policy/dynamic_tf_policy.py", line 198, in __init__
    before_loss_init(self, obs_space, action_space, config)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/policy/tf_policy_template.py", line 127, in before_loss_init_wrapper
    self._extra_action_fetches = extra_action_fetches_fn(self)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/agents/ppo/ppo_policy.py", line 170, in vf_preds_and_logits_fetches
    SampleBatch.VF_PREDS: policy.model.value_function(),
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/models/tf/modelv1_compat.py", line 148, in value_function
    seq_lens=None)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/models/catalog.py", line 481, in get_model
    seq_lens)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/models/catalog.py", line 521, in _get_model
    num_outputs, options)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/models/model.py", line 57, in __init__
    input_dict["obs"], obs_space)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/models/model.py", line 230, in restore_original_dimensions
    return _unpack_obs(obs, obs_space.original_space, tensorlib=tensorlib)
  File "/afs/ece.cmu.edu/usr/charlieh/.local/lib/python3.6/site-packages/ray/rllib/models/model.py", line 260, in _unpack_obs
    prep.shape[0], obs.shape))
ValueError: Expected flattened obs shape of [None, 143], got (?, 256)
`
My state space is
`obs_space = spaces.Tuple(spaces.Discrete(35),spaces.Discrete(35),spaces.Discrete(35), spaces.Discrete(3), spaces.Discrete(35))`
and if I run 
`prep = get_preprocessor(spy_state_space)(spy_state_space)
print(prep)`
I get

> (143,)

As expected.  I believe that this is similar to issue #3367 (except this time the observation space is a Tuple rather than a Dict.  Am I doing something wrong or is this a bug?  If requested, I can also try posting my code.

### PR DESCRIPTION
**Pull request by sven1977**
Fixes issue #6884 

#6884

closes #6884

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: rllib/models/tf/modelv1_compat.py [hunk 1]
```diff
@@ -124,18 +123,21 @@ def value_function(self):
...
+                        logger.warning(
+                            "It is not recommended to use an LSTM model "
+                            "with the `vf_share_layers=False` option. "
+                            "If you do not want to use a shared policy+vf "
+                            "network, you can implement a custom LSTM "
+                            "model that overrides the value_function() "
+                            "method. "
+                            "`use_lstm` has been set to False for the "
+                            "vf-part of your network.")
+                        input_dict["obs"] = np.array([self.obs_space.sample()])
```
Comment (ericl) [inline rllib/models/tf/modelv1_compat.py]: Isn't this not a tensor?

Comment (sven1977) [reply rllib/models/tf/modelv1_compat.py]: Yes, but this is only for the first dummy pass, correct?

Comment (sven1977) [reply rllib/models/tf/modelv1_compat.py]: You are right. I was thinking tf2.x :/
This will screw up the static graph. I'll take another look.

Comment (sven1977) [reply rllib/models/tf/modelv1_compat.py]: This is resolved now: vf-NN takes LSTM's last output as input (same as for non-Tuple observations). The problem was the preprocessor/shape check, which failed. This check only happened for Tuple/Dict observations, but is not necessary in the case of a shared vf (b/c no preprocessing of the LSTM output necessary).

File: rllib/examples/random_env.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+from gym.spaces import Tuple, Discrete
+import numpy as np
+
+from ray.rllib.agents.ppo import PPOTrainer
+from ray.rllib.utils import try_import_tf
+
+tf = try_import_tf()
+
+
+class RandomEnv(gym.Env):
```
Comment (ericl) [inline rllib/examples/random_env.py]: Should this be a test? Btw, we can also add an entry to `test_supported_spaces.py` which historically has been used to catch these sort of bugs.

**Review by ericl — APPROVED** [review]

File: rllib/examples/random_env.py (hunk 2 above)
Comment (sven1977) [reply rllib/examples/random_env.py]: Yes, would be useful to add as a test.
I'll add both (Radom-env and supported-spaces) as tests (to Jenkins AND the new BAZEL branch).

Comment (sven1977) [reply rllib/examples/random_env.py]: Hmm, we do check the support for tuple spaces for PPO, but not with the special config combination that caused this issue. Yeah, I was thinking about that last night: We should ideally expand this idea to more settings (add more loops): obs-space, action-space, framework, algo, lstm (yes/no), num_workers (0, 1, or 2), etc.. That would increase our test coverage a lot.

**Comment by sven1977** [comment]
@ericl: tests are ok now.

**Pull request merged**